Barotrauma/BarotraumaShared/SharedSource/Items/Components/Machines/Sonar.cs

```diff
@@ -183,73 +192,72 @@ public override void Update(float deltaTime, Camera cam)
-                            item.AiTarget.SectorDir = new Vector2(pingDirection.X, -pingDirection.Y);
-                        }
+                            Radius = 0.0f,
+                            PrevRadius = 0.0f,
+                            IsDirectional = useDirectionalPing,
+                            Direction = pingDirection
+                        };
+                        activePings.Add(newPing);
+
                         item.Use(deltaTime);
+                        pingTimer = 0.0f;
                     }
+
+                    aiPingCheckPending = true;
+                    pingTimer += deltaTime * (1 / PingInterval);
                 }
                 else
                 {
                     aiPingCheckPending = false;
+                    pingTimer = 1.0f;
                 }
             }
+            else
+            {
+                aiPingCheckPending = false;
+                activePings.Clear();
+                pingTimer = 1.0f;
+            }
 
-            for (var pingIndex = 0; pingIndex < activePingsCount;)
+            if (activePings.Any())
             {
-                if (item.AiTarget != null)
-                {
-                    float range = MathUtils.InverseLerp(item.AiTarget.MinSoundRange, item.AiTarget.MaxSoundRange, Range * activePings[pingIndex].State / zoom);
-                    item.AiTarget.SoundRange = Math.Max(item.AiTarget.SoundRange, MathHelper.Lerp(item.AiTarget.MinSoundRange, item.AiTarget.MaxSoundRange, range));
-                }
-                if (activePings[pingIndex].State > 1.0f)
+                for (var pingIndex = 0; pingIndex < activePings.Count; pingIndex++)
                 {
-                    var lastIndex = --activePingsCount;
-                    var oldActivePing = activePings[pingIndex];
-                    activePings[pingIndex] = activePings[lastIndex];
-                    activePings[lastIndex] = oldActivePing;
-                    if (currentPingIndex == lastIndex)
+                    ActivePing activePing = activePings[pingIndex];
+                    if (activePing.Radius > Range)
                     {
-                        currentPingIndex = pingIndex;
+                        activePings.RemoveAt(pingIndex);
+                    }
+                    else
+                    {
+                        activePing.Radius += deltaTime * PingSpeed;
+                    }
+
+                    if (item.AiTarget != null)
+                    {
+                        item.AiTarget.SoundRange = Math.Max(item.AiTarget.SoundRange, MathHelper.Lerp(item.AiTarget.MinSoundRange, item.AiTarget.MaxSoundRange, activePing.Radius / Range));
                     }
                 }
-                else
+
+                // TODO: Allow multiple seperated sectors & iterate over each active ping to disallow pinging a sector then moving the ping direction to gain information without aggro'ing enemies.
```

Comment:
Unaddressed todo note. Now the sector seems to be based on the latest ping, which could potentially be abused.